aptos-move/move-examples/dao/nft_dao/sources/nft_dao.move

```diff
@@ -0,0 +1,1141 @@
+        while (i < vector::length(&token_names)) {
+            let token_name = *vector::borrow(&token_names, i);
+            let property_version = *vector::borrow(&property_versions, i);
+            let token_id = token::create_token_id_raw(gtoken.creator, gtoken.collection, token_name, property_version);
+            // check if this token already voted
+            assert!(!bucket_table::contains(&stats.no_votes, &token_id), error::invalid_argument(ETOKEN_ALREADY_VOTED));
+            assert!(!bucket_table::contains(&stats.yes_votes, &token_id), error::invalid_argument(ETOKEN_ALREADY_VOTED));
+
+            // this account owns the token
+            assert!(token::balance_of(signer::address_of(account), token_id) == 1, error::permission_denied(ENOT_OWN_THE_VOTING_DAO_TOKEN));
+            if (vote) {
+                stats.total_yes = stats.total_yes + 1;
+                bucket_table::add(&mut stats.yes_votes, token_id, voter_addr);
+            } else {
+                stats.total_no = stats.total_no + 1;
+                bucket_table::add(&mut stats.no_votes, token_id, voter_addr);
+            };
+            i = i + 1;
+        };
+
+        nft_dao_events::emit_voting_event(
+            voter_addr,
+            nft_dao,
+            proposal_id,
+            vote,
+            token_names,
+            property_versions,
+        );
+    }
+
+    /// Anyone can call the resolve function to resolve a proposal.
+    public entry fun resolve<CoinType>(proposal_id: u64, nft_dao: address) acquires Proposals, DAO, ProposalVotingStatistics {
+        let dao = borrow_global<DAO>(nft_dao);
+        // assert the proposal voting ended
+        let proposals = borrow_global<Proposals>(nft_dao);
+        let proposal = table::borrow(&proposals.proposals, proposal_id);
+        let now = timestamp::now_seconds();
+        assert!(now >= proposal.start_time_sec + dao.voting_duration, error::invalid_argument(EPROPOSAL_ENDED));
+        // assert the proposal is unresolved yet
+        assert!(proposal.resolution == PROPOSAL_PENDING, error::invalid_argument(EPROPOSAL_RESOLVED));
+        resolve_internal<CoinType>(option::none(), proposal_id, nft_dao);
+    }
+
+    /// Admin can veto an active proposal
+    public entry fun admin_veto_proposal(admin: &signer, proposal_id: u64, nft_dao: address) acquires DAO, Proposals {
+        let dao = borrow_global_mut<DAO>(nft_dao);
+        assert!(dao.admin == signer::address_of(admin), error::permission_denied(EINVALID_ADMIN_ACCOUNT));
+        // assert the proposal is still active
+        let proposals = borrow_global_mut<Proposals>(nft_dao);
+        let proposal = table::borrow_mut(&mut proposals.proposals, proposal_id);
+        // assert the proposal is unresolved yet
+        assert!(proposal.resolution == PROPOSAL_PENDING, error::invalid_argument(EPROPOSAL_RESOLVED));
+        proposal.resolution = PROPOSAL_VETOED_BY_ADMIN;
+
+        nft_dao_events::emit_resolve_event(
```

Comment:
Do we want a different event here for veto?

Reply:
or maybe the reason why it was resolved

Reply:
added new events for admin veto and resolve. previously, we can distinguish admin veto/resolve through `proposal.resolution` field. But I guess it is not very obviously for a lot of users.